cpp/tests/nvcategory/nvcategory_test.cu

```diff
@@ -0,0 +1,711 @@
+
+#include "gtest/gtest.h"
+
+#include <cudf.h>
+#include <utilities/cudf_utils.h>
+
+#include <cudf/functions.h>
+#include <cudf/types.h>
+#include <iostream>
+#include <random>
+
+#include <NVCategory.h>
+#include <NVStrings.h>
+#include "rmm/rmm.h"
+#include <cstring>
+#include "tests/utilities/cudf_test_utils.cuh"
+#include "tests/utilities/cudf_test_fixtures.h"
+#include "bitmask/bit_mask.h"
+
+// See this header for all of the handling of valids' vectors 
+#include "tests/utilities/valid_vectors.h"
+
+namespace {
+std::string const default_chars = 
+	"abcdefghijklmnaoqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890";
+}
+
```

Comment:
Make sure to test any new error conditions added in the PR that may be returned that aren't already tested. e.g. are there any new conditions under which GDF_COLUMN_TYPE_MISMATCH, COLUMN_SIZE_MISMATCH, or GDF_NOT_IMPLEMENTED will be returned?

Reply:
Are you saying you want a test case for every  error code that coudl be thrown by the code? While I am not averse to this in principle this seems like a MUCH higher bar than anything I  have seen in the rest of the code. We have tons of code that isn't covered at all.

Reply:
It's not higher than the bar I set for myself. See the tests added in PR #913, for example. They test all expected error output situations of each function that is added.  Basically for every `GDF_REQUIRE`, I have a corresponding test that intentionally violates the requirement.